runtime/sema/type.go

```diff
@@ -3779,6 +3827,137 @@ func (t *CompositeType) GetNestedTypes() *StringTypeOrderedMap {
 	return t.NestedTypes
 }
 
+func (t *CompositeType) isTypeIndexableType() bool {
+	// resources and structs only can be indexed for attachments
+	return t.Kind == common.CompositeKindResource || t.Kind == common.CompositeKindStructure
+}
+
+func (t *CompositeType) TypeIndexingElementType(indexingType Type) Type {
+	return &OptionalType{
+		&ReferenceType{
+			Type: indexingType,
+		},
+	}
+}
+
+func (t *CompositeType) IsValidIndexingType(ty Type) bool {
+	attachmentType, isComposite := ty.(*CompositeType)
+	return isComposite &&
+		IsSubType(t, attachmentType.baseType) &&
+		attachmentType.IsResourceType() == t.IsResourceType()
+}
+
+const compositeForEachAttachmentFunctionName = "forEachAttachment"
+
+const compositeForEachAttachmentFunctionDocString = `
+Iterates over the attachments present on the receiver, applying the function argument to each.
+The order of iteration is undefined. If a type argument is provided, only attachments that conform
+to the specified interface are iterated on, the others are filtered out. 
```

Comment:
should we have the function return a `Bool` like the other `forEach` functions? that way we have a more unified API in case we decide to generalize `forEach` to any collection type in the future